cmd/oras/root/manifest/index/create.go

```diff
@@ -0,0 +1,207 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package index
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/opencontainers/image-spec/specs-go"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/spf13/cobra"
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/errdef"
+	"oras.land/oras/cmd/oras/internal/command"
+	"oras.land/oras/cmd/oras/internal/display"
+	oerrors "oras.land/oras/cmd/oras/internal/errors"
+	"oras.land/oras/cmd/oras/internal/option"
+	"oras.land/oras/cmd/oras/internal/output"
+	"oras.land/oras/internal/contentutil"
+	"oras.land/oras/internal/descriptor"
+	"oras.land/oras/internal/listener"
+)
+
+type createOptions struct {
+	option.Common
+	option.Target
+
+	sources   []string
+	extraRefs []string
+}
+
+func createCmd() *cobra.Command {
+	var opts createOptions
+	cmd := &cobra.Command{
+		Use:   "create [flags] <name>[:<tag[,<tag>][...]] [{<tag>|<digest>}...]",
+		Short: "Create and push an index from provided manifests",
+		Long: `Create and push an index to a repository or an OCI image layout
+
+Example - create an index from source manifests tagged amd64, arm64, darwin in the repository
+ localhost:5000/hello, and push the index without tagging it:
+  oras manifest index create localhost:5000/hello amd64 arm64 darwin
+
+Example - create an index from source manifests tagged amd64, arm64, darwin in the repository
+ localhost:5000/hello, and push the index with tag 'latest':
+  oras manifest index create localhost:5000/hello:latest amd64 arm64 darwin
+
+Example - create an index from source manifests using both tags and digests, 
+ and push the index with tag 'latest':
+  oras manifest index create localhost:5000/hello latest amd64 sha256:xxx darwin
+
+Example - create an index and push it with multiple tags:
+  oras manifest index create localhost:5000/tag1, tag2, tag3 amd64 arm64 sha256:xxx
+`,
+		Args: cobra.MinimumNArgs(1),
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			refs := strings.Split(args[0], ",")
+			opts.RawReference = refs[0]
+			opts.extraRefs = refs[1:]
+			opts.sources = args[1:]
+			return option.Parse(cmd, &opts)
+		},
+		Aliases: []string{"pack"},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return createIndex(cmd, opts)
+		},
+	}
+	option.ApplyFlags(&opts, cmd.Flags())
+	return oerrors.Command(cmd, &opts.Target)
+}
+
+func createIndex(cmd *cobra.Command, opts createOptions) error {
+	ctx, logger := command.GetLogger(cmd, &opts.Common)
+	target, err := opts.NewTarget(opts.Common, logger)
+	if err != nil {
+		return err
+	}
+	// if a digest is given as the index reference, we need to ignore it
+	opts.RawReference, _, _ = strings.Cut(opts.RawReference, "@")
```

Comment:
If the user gives a digest for index reference, should we silently ignore it, or return an error and fail?

Reply:
Nested indexes are supported in OCI and should be a test case as well.

Reply:
added a test case for creating nested indexes.

Reply:
If user gives an invalid tag, we should return an error.